lib/evmone/lru_cache.hpp

```diff
@@ -0,0 +1,157 @@
+    /// any performance difference.
+    Map map_;
+
+    /// Marks an element as the most recently used by moving it to the back of the LRU list.
+    void move_to_back(LRUIterator it) noexcept { lru_list_.splice(lru_list_.end(), lru_list_, it); }
+
+public:
+    /// Constructs the LRU cache with the given capacity.
+    ///
+    /// @param capacity  The fixed capacity of the cache.
+    explicit LRUCache(size_t capacity) : capacity_(capacity)
+    {
+        // Reserve map to the full capacity to prevent any rehashing.
+        map_.reserve(capacity);
+    }
+
+    /// Clears the cache by deleting all the entries.
+    void clear() noexcept
+    {
+        map_.clear();
+        lru_list_.clear();
+    }
+
+
+    /// Retrieves the copy of the value associated with the specified key.
+    ///
+    /// @param key  The key of the entry to retrieve.
+    /// @return     An optional containing the copy of the value if the key is found,
+    ///             or an empty optional if not.
+    std::optional<Value> get(const Key& key) noexcept
+    {
+        if (const auto it = map_.find(key); it != map_.end())
+        {
+            move_to_back(it->second);
+            return it->second->value;
+        }
+        return {};
+    }
+
+    /// Inserts or updates the value associated with the specified key.
+    ///
+    /// @param key    The key of the entry to insert or update.
+    /// @param value  The value to associate with the key.
+    void put(Key key, Value value)
+    {
+        // Implementation is split into two variants: cache full or not.
+        // Once the cache is full, its size never shrinks therefore from now on this variant is
+        // always executed.
+
+        if (map_.size() == capacity_)
+        {
+            // When the cache is full we avoid erase-emplace pattern by using the map's node API.
+
+            using namespace std;  // for swap usage with ADL
+
+            // Get the least recently used element.
+            auto lru_it = lru_list_.begin();
+
+            // Extract the map node with the to-be-evicted element and reuse it for the new
+            // key-value pair. This makes the operation allocation-free.
+            auto node = map_.extract(lru_it->key);
+            swap(node.key(), key);
+            if (auto [it, inserted, node2] = map_.insert(std::move(node)); !inserted)
+            {
+                // Failed re-insertion means the element with the new key is already in the cache.
+                // Rollback the eviction by re-inserting the node with original key back.
+                swap(key, node2.key());
+                map_.insert(std::move(node2));
+
+                // Returned iterator points to the element matching the key
+                // which value must be updated.
+                lru_it = it->second;
+            }
+            lru_it->value = std::move(value);  // Replace/update the value.
+            move_to_back(lru_it);
```

Comment:
Juggling of keys is quite confusing, I'd maybe add more comments like
```suggestion
            auto node = map_.extract(lru_it->key); // node.key() is LRU key
            swap(node.key(), key); // node.key() is new key, key is LRU key
            if (auto [it, inserted, node2] = map_.insert(std::move(node)); !inserted)
            {
                // node2.key() is new key (already in cache)
                // Failed re-insertion means the element with the new key is already in the cache.
                // Rollback the eviction by re-inserting the node with original key back.
                swap(key, node2.key()); // key is new key (already in cache), node2.key() is LRU key
                map_.insert(std::move(node2));

                // Returned iterator points to the element matching the key
                // which value must be updated.
                lru_it = it->second;
            }
            lru_it->value = std::move(value);  // Replace/update the value.
            move_to_back(lru_it);
```

Reply:
Yeah, this value update for existing key feature is quite annoying to implement. And in many cases it is never used... I will try to improve the comments.